core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -44,126 +44,355 @@
         boolean isNewTransaction = false;
-        boolean rolledBack = false;
-        V result = null;
+        boolean isRolledBack = false;
+        boolean isCommitted = false;
         try {
             DefaultTransactionDefinition definition = new DefaultTransactionDefinition(transactionRequirement.propagationId);
             definition.setName(name);
             definition.setReadOnly(readOnly);
-            transaction = transactionManager.getTransaction(definition);
-            isNewTransaction = transaction.isNewTransaction();
+            status = transactionManager.getTransaction(definition);
+            isNewTransaction = status.isNewTransaction();
 
             if (isNewTransaction) {
-                Collection<NewTransactionListener> listeners = context.getBeansOfType(NewTransactionListener.class).values();
-                for (NewTransactionListener listener : listeners) {
-                    listener.newTransaction(transaction);
+                try {
+                    fireNewTransaction(context, transactionManager, status);
+                }
+                catch(Throwable t) {
+                    Log.warning(Log.JEEVES, "New transaction:", t);
+                    // warning as we continue with action below
                 }
             }
 
-            result = action.doInTransaction(transaction);
+            return action.doInTransaction(status);
 
-        } catch (Throwable e) {
-            Log.error(Log.JEEVES, "Error occurred within a transaction", e);
-            if (exception[0] == null) {
-                exception[0] = e;
+        } catch (Throwable exception) {
+            Log.error(Log.JEEVES, "Error occurred within a transaction", exception);
+            isRolledBack = rollbackIfNotRolledBack(context, transactionManager, status, isRolledBack);
+            if (exception instanceof RuntimeException) {
+                throw (RuntimeException) exception;
+            } else if (exception instanceof Error) {
+                throw (Error) exception;
+            } else {
+                throw new RuntimeException("Run in transaction '"+action+"': "+exception.getLocalizedMessage(), exception);
             }
-            rolledBack = true;
-            doRollback(context, transactionManager, transaction);
         } finally {
             try {
                 if (readOnly) {
-                    doRollback(context, transactionManager, transaction);
-                } else if (!rolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
-                    doCommit(context, transactionManager, transaction);
+                    isRolledBack = rollbackIfNotRolledBack(context, transactionManager, status, isRolledBack);
+                } else if (!isRolledBack && (isNewTransaction || commitBehavior == CommitBehavior.ALWAYS_COMMIT)) {
+                    isCommitted = true;
+                    doCommit(context, transactionManager, status);
```

Comment:
This logic does not look correct:
1. If `doCommit` fails (and it will) the flag `isCommitted` is already set indicating it has succeeded.
2. The catch statement below will try and rollback
3. The finally log will indicate that the transaction is both committed and rollback